test/model/typing_status_test.dart

```diff
@@ -0,0 +1,223 @@
+      checkTypists({dmNarrow: [eg.otherUser], groupNarrow: [eg.thirdUser]});
+      checkNotifiedOnce();
+    });
+
+    test('add a typist in the same narrow', () {
+      prepareModel();
+
+      handleTypingEvent(groupNarrow, TypingOp.start, eg.otherUser);
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+
+      handleTypingEvent(groupNarrow, TypingOp.start, eg.thirdUser);
+      checkTypists({groupNarrow: [eg.otherUser, eg.thirdUser]});
+      checkNotifiedOnce();
+    });
+
+    test('sort dm recipients', () {
+      prepareModel(selfUserId: 5);
+      final recipientIds = [5, 4, 10, 8, 2, 1];
+
+      final eventUnsorted = TypingEvent(id: 1, op: TypingOp.start,
+        senderId: 1,
+        messageType: MessageType.direct,
+        recipientIds: recipientIds,
+        streamId: null, topic: null);
+      // DmNarrow's constructor expects the recipient IDs to be sorted,
+      // and [model.handleTypingEvent] should handle that.
+      model.handleTypingEvent(eventUnsorted);
+      check(model.typistIdsInNarrow(
+        DmNarrow(allRecipientIds: recipientIds..sort(), selfUserId: 5),
+      )).single.equals(1);
+    });
+  });
+
+  group('handle typing stop events', () {
+    test('remove a typist from an unknown narrow', () {
+      prepareModel();
+
+      handleTypingEvent(groupNarrow, TypingOp.start, eg.otherUser);
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+
+      handleTypingEvent(dmNarrow, TypingOp.stop, eg.otherUser);
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotNotified();
+    });
+
+    test('remove one of two typists in the same narrow', () {
+      prepareModel(typistsByNarrow: {
+        groupNarrow: [eg.otherUser, eg.thirdUser],
+      });
+
+      handleTypingEvent(groupNarrow, TypingOp.stop, eg.otherUser);
+      checkTypists({groupNarrow: [eg.thirdUser]});
+      checkNotifiedOnce();
+    });
+
+    test('remove all typists in a narrow', () {
+      prepareModel(typistsByNarrow: {dmNarrow: [eg.otherUser]});
+
+      handleTypingEvent(dmNarrow, TypingOp.stop, eg.otherUser);
+      checkTypists({});
+      checkNotifiedOnce();
+    });
+
+    test('remove typists from different narrows', () {
+      prepareModel(typistsByNarrow: {
+        dmNarrow: [eg.otherUser],
+        groupNarrow: [eg.thirdUser],
+        topicNarrow: [eg.fourthUser],
+      });
+
+      handleTypingEvent(groupNarrow, TypingOp.stop, eg.thirdUser);
+      checkTypists({dmNarrow: [eg.otherUser], topicNarrow: [eg.fourthUser]});
+      checkNotifiedOnce();
+
+      handleTypingEvent(dmNarrow, TypingOp.stop, eg.otherUser);
+      checkTypists({topicNarrow: [eg.fourthUser]});
+      checkNotifiedOnce();
+
+      handleTypingEvent(topicNarrow, TypingOp.stop, eg.fourthUser);
+      checkTypists({});
+      checkNotifiedOnce();
+    });
+  });
+
+  group('cancelling old timer', () {
+    test('when typing stopped early', () => awaitFakeAsync((async) async {
```

Comment:
This group doesn't feel to me like the logical way to organize these two tests. Both test cases are good ones to have; but this one, let's put under the "stop events" group, with name "cancel timer".

The other case below, let's put under the "start events" group, with name "cancel existing timer".

In particular, these don't feel to me like two cases of one idea — it's more like two separate pieces of logic, that are part of what's expected for a stop event and a start event respectively.

Reply:
Or perhaps better: these two could be organized together with the cases currently under "expiry period" below, as a group that's about the timers.

Then the narrative, test case by test case, could be:
* start sets a timer — look, just wait 15 seconds and the typist gets removed (the current "typist is removed when the expiry period ends")
* stop cancels the timer (the current "when typing stopped early")
* repeated start resets the timer — look, there's just one timer pending, and you have to wait 15 seconds before the typist gets removed, even if it'd already been 10 seconds since the original start (the current "repeated typing start event resets the timer", plus the `pendingTimers` checks from the current "when typing repeatedly started")

I think that makes a pretty clean story.